valet/pathproc.go

```diff
@@ -27,61 +27,72 @@ import (
 
 	logs "github.com/kjsanger/logshim"
 	"github.com/pkg/errors"
-
-	"github.com/kjsanger/valet/utilities"
 )
 
 type token struct{}
 type semaphore chan token
 
 type ProcessParams struct {
-	Root          string        // The local root directory to work on
-	MatchFunc     FilePredicate // The file selecting predicate
-	PruneFunc     FilePredicate // The local directory tree pruning predicate
-	Plan          WorkPlan      // The plan for selected files
-	SweepInterval time.Duration // The interval between sweeps of the local directory tree
-	MaxProc       int           // The maximum number of threads to run
+	Root          string        // The local root directory to work on.
+	MatchFunc     FilePredicate // The file selecting predicate.
+	PruneFunc     FilePredicate // The local directory tree pruning predicate.
+	Plan          WorkPlan      // The plan for selected files.
+	SweepInterval time.Duration // The interval between sweeps of the local directory tree.
+	MaxProc       int           // The maximum number of threads to run.
 }
 
+// ProcessFiles detects files to work on, dispatches any files found to
+// suitable work functions and monitors any errors that occur during the
+// detection and processing steps. The function will continue to run until
+// cancelled.
+//
+// Errors that occur in detection are logged as warnings, but do not cause this
+// function to return an error itself. Error that occur during processing are
+// counted. If when cancelled, this function has counted any processing errors,
+// it will return an error itself.
 func ProcessFiles(cancelCtx context.Context, params ProcessParams) error {
 
 	wpaths, werrs := WatchFiles(cancelCtx, params.Root, params.MatchFunc,
 		params.PruneFunc)
 	fpaths, ferrs := FindFilesInterval(cancelCtx, params.Root, params.MatchFunc,
 		params.PruneFunc, params.SweepInterval)
 
-	mpaths := MergeFileChannels(wpaths, fpaths)
-	merrs := MergeErrorChannels(werrs, ferrs)
-	done := make(chan token)
-
+	paths := MergeFileChannels(wpaths, fpaths)
+	errs := MergeErrorChannels(werrs, ferrs)
 	log := logs.GetLogger()
-	var errs []error
 
+	// Inform the user that cancellation has started because it can take a
+	// while for jobs to complete. This blocks until then.
 	go func() {
-		defer func() { done <- token{} }()
-
-		errs = append(errs, DoProcessFiles(mpaths, params.Plan, params.MaxProc))
-	}()
-
-done:
-	for {
 		select {
-		case <-done:
-			log.Info().Msg("processing done")
-			break done
 		case <-cancelCtx.Done():
```

Comment:
This doesn't need to be in a select any more.

Reply:
Cancellation (e.g. following trapped `SIGINT`) is _supposed_ to be the only way `DoProcessFiles` can return, so this goroutine _should_ always return. However, it seems wise to have another way of stopping the potential leak in case that changes. I retained the select and added another branch to return if `DoProcessFiles`  completes without cancelling.